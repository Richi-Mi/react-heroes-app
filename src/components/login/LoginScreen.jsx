import React, { useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import AuthContext from '../../auth/authContext';
import useForm from '../../hooks/useForm';
import types from '../../types/types';

import './../../design/login.css'
import './../../design/search.css'

const LoginScreen = () => {
    const navigate = useNavigate();
    const { dispatch } = useContext( AuthContext );
    const [ values, handleInputChange, reset ] = useForm({
        username: ''
    });
    const { username } = values 

    const handleLogin = () => {
        const lastPath = localStorage.getItem('lastPath') || '/';
        const action = {
            type: types.login,
            payload: {
                name: username
            }
        }
        dispatch( action );

        reset()

        navigate( lastPath, { 
            replace: true
        });
    }
    
    return (
        <div className='container_login'>
            <h1> Login Screen </h1>
            <hr />
            <div className="form_search">
                <input 
                    type='text'
                    className='form_input'
                    placeholder='Your Name'
                    name='username'
                    autoComplete='off'
                    value={ username }
                    onChange={ handleInputChange }
                />
                <button className='btn_search btn_login' onClick={ handleLogin }> Login </button>   
            </div>         
        </div>
    )
};

export default LoginScreen;
