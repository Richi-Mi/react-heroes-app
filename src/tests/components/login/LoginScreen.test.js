import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import AuthContext from '../../../auth/authContext';
import LoginScreen from '../../../components/login/LoginScreen';
import types from '../../../types/types';

const mockDispatch = jest.fn();
const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useNavigate: () => mockNavigate
}))

describe(' * Pruebas en el componente <LoginScreen />', () => {
	
	const value = {
		dispatch: mockDispatch,
		user: {
			logged: false
		}
	}

	const wrappLog = mount(
		<AuthContext.Provider value={ value }>
			<MemoryRouter initialEntries={ ['/login'] }>
				<LoginScreen />
			</MemoryRouter>
		</AuthContext.Provider>
	);

	beforeEach( () => {
		jest.clearAllMocks();
	} )

    test(' - Debe mostrarse correctamente', () => {

		expect( wrappLog ).toMatchSnapshot();

    });

    test(' - Debe de llamar el dispatch y la navegación', () => {
		const nameUser = 'Ricardo'

		wrappLog.find('.form_input').simulate('change', {
			target: {
				name: 'username',
				value: nameUser
			}
		});

		wrappLog.find('.btn_search').simulate('click');

		// < Dispatch
		expect( mockDispatch ).toHaveBeenCalled();
		expect( mockDispatch ).toHaveBeenCalledTimes(1);
		expect( mockDispatch ).toHaveBeenCalledWith({
            type: types.login,
        	payload: {
            	name: nameUser
        	}
		});
		// Navigate
		expect( mockNavigate ).toHaveBeenCalled();
		expect( mockNavigate ).toHaveBeenCalledTimes(1);
		expect( mockNavigate ).toHaveBeenCalledWith('/', {
			replace: true
		});

    });
    
	test(' - Debe de llamar el dispatch y la navegación con el ultimo path usado', () => {
		const nameUser = 'Ricardo';
		const lastPath = '/dc';

		localStorage.setItem('lastPath', lastPath)

		wrappLog.find('.form_input').simulate('change', {
			target: {
				name: 'username',
				value: nameUser
			}
		});

		wrappLog.find('.btn_search').simulate('click');

		// < Dispatch
		expect( mockDispatch ).toHaveBeenCalled();
		expect( mockDispatch ).toHaveBeenCalledTimes(1);
		expect( mockDispatch ).toHaveBeenCalledWith({
            type: types.login,
        	payload: {
            	name: nameUser
        	}
		});
		// Navigate
		expect( mockNavigate ).toHaveBeenCalled();
		expect( mockNavigate ).toHaveBeenCalledTimes(1);
		expect( mockNavigate ).toHaveBeenCalledWith( lastPath, {
			replace: true
		});
	});
	
    
});
