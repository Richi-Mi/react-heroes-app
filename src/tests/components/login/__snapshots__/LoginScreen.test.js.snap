// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` * Pruebas en el componente <LoginScreen />  - Debe mostrarse correctamente 1`] = `
<div
  className="container_login"
>
  <h1>
     Login Screen 
  </h1>
  <hr />
  <div
    className="form_search"
  >
    <input
      autoComplete="off"
      className="form_input"
      name="username"
      onChange={[Function]}
      placeholder="Your Name"
      type="text"
      value=""
    />
    <button
      className="btn_search btn_login"
      onClick={[Function]}
    >
       Login 
    </button>
  </div>
</div>
`;
